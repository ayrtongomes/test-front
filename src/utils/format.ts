export function formatCardNum(value: string): string {
  const formatted = value.replace(/\D/g, '') // remove non-digit characters
  const groups = formatted.match(/.{1,4}/g) // split into groups of 4 digits
  return groups ? groups.join(' ') : formatted
}

export function formatCardExpirationDate(value: string): string {
  const v = value.replace(/\D/g, '')
  const month = v.slice(0, 2)
  const year = v.slice(2, 4)
  return `${month}/${year}`
}
