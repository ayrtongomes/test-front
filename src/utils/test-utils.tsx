import { render, RenderOptions } from '@testing-library/react'
import { rest } from 'msw'
import { ReactElement, ReactNode } from 'react'
import { ThemeProvider } from 'styled-components'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

import { theme } from '../styles/theme'
import { mockData } from '../test/mocks'

type RenderThemeProps = Omit<RenderOptions, 'queries'>

const renderTheme = (children: ReactElement, { ...renderOptions }: RenderThemeProps = {}) =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>, renderOptions)

export * from '@testing-library/react'
export { renderTheme as render }

export const handlers = [
  rest.get('*/data/*', (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        ...mockData
      })
    )
  })
]

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        refetchOnMount: false,
        refetchIntervalInBackground: false,
        refetchOnWindowFocus: false,
        refetchInterval: false,
        refetchOnReconnect: false
      }
    }
  })

export function renderWithClient(ui: ReactNode) {
  const testQueryClient = createTestQueryClient()
  const { rerender, ...result } = renderTheme(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
  )
  return {
    ...result,
    rerender: (rerenderUi: ReactNode) =>
      rerender(<QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>)
  }
}

export function createWrapper() {
  const testQueryClient = createTestQueryClient()
  return ({ children }: any) => (
    <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
  )
}
