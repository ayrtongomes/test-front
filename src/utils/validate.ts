export function isExpirationDateValid(value: string): boolean {
  const [month, year] = value.split('/').map((str) => parseInt(str, 10))
  const now = new Date()
  const currentYear = now.getFullYear() % 100
  const currentMonth = now.getMonth() + 1

  if (year < currentYear || (year === currentYear && month < currentMonth)) {
    return false
  }

  return true
}

export function isNameValid(value: string): boolean {
  const names = value.split(' ')
  const regex = /^[a-zA-Z ]*$/

  if (!regex.test(value)) {
    return false
  }

  if (names.length < 2) {
    return false
  }

  const lastName = names[names.length - 1]

  if (lastName.length <= 2 || /^[A-Z][a-z]*\.$/.test(lastName)) {
    return false
  }

  return true
}
