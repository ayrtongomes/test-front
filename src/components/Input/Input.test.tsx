import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { render } from '../../utils/test-utils'

import { Input } from '.'

jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  useFormContext: jest.fn().mockReturnValue({
    register: () => {
      return {
        name: 'test', // assign name prop
        onChange: jest.fn() // assign onChange event
      }
    },
    formState: { errors: { test: { message: 'Error message' } } }
  })
}))

describe('Input component', () => {
  it('renders with placeholder', () => {
    render(<Input name="test" placeholder="placeholder text" onChange={() => {}} label={'Label'} />)

    expect(screen.getByPlaceholderText('placeholder text')).toBeInTheDocument()
  })

  it('changes value when typing', async () => {
    const onInputChange = jest.fn()
    render(<Input onChange={onInputChange} label="Label" name="Label" />)

    const input = screen.getByRole('textbox')
    const text = 'text changed'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })
  })

  it('Renders with error', () => {
    const { container } = render(<Input name="test" label="Label" />)

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
