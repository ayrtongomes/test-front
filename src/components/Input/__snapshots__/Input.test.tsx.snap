// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Renders with error 1`] = `
.c2 {
  background: #FFFFFF;
  border: 1px solid #CCCCCC !important;
  border-radius: 4px;
  border: none;
  color: #000000;
  font-size: 14pxpx;
  margin-top: 8px;
  outline: none;
  padding: 14px;
  width: 100%;
  height: 44px;
}

.c2:focus {
  border-color: #000000 !important;
}

.c1 {
  font-size: 16px;
  line-height: 20px;
  color: #515151;
}

.c0 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  margin-bottom: 10px;
  margin-top: 10px;
  position: relative;
  width: 100%;
}

.c0 + .input {
  margin-left: 8px;
}

.c3 {
  color: #DE1616;
  font-size: 12px;
  margin-top: 8px;
}

<div
  class="c0 input"
  width="100%"
>
  <label
    class="c1"
    for="test"
  >
    Label
  </label>
  <input
    class="c2"
    id="test"
    maxlength="1000"
    minlength="0"
    name="test"
  />
  <span
    class="c3"
  >
    Error message
  </span>
</div>
`;
